plutus-emulator/src/Wallet/Emulator/NodeClient.hs

```diff
@@ -0,0 +1,150 @@
+    _txPool           :: TxPool, -- ^ The pool of pending transactions.
+    _index            :: Index.UtxoIndex -- ^ The UTxO index, used for validation.
+} deriving (Show)
+
+makeLenses ''NodeClientState
+
+-- | Validate a block given the current slot and UTxO index, returning the valid
+--   transactions, success/failure events, remaining transactions and the
+--   updated UTxO set.
+validateBlock :: Slot -> Index.UtxoIndex -> [Tx] -> (ValidatedBlock, Index.UtxoIndex)
+validateBlock slot@(Slot s) idx txns =
+    let
+        -- Select those transactions that can be validated in the
+        -- current slot
+        (eligibleTxns, rest) = partition (canValidateNow slot) txns
+
+        -- Validate eligible transactions, updating the UTXO index each time
+        (processed, idx') =
+            flip runState idx $ for eligibleTxns $ \t -> do
+                r <- validateEm slot t
+                pure (t, r)
+
+        -- The new block contains all transaction that were validated
+        -- successfully
+        block = fst <$> filter (isNothing . snd) processed
+
+        -- Also return an `EmulatorEvent` for each transaction that was
+        -- processed
+        nextSlot = Slot (s + 1)
+        events   = SlotAdd nextSlot : (uncurry mkEvent <$> processed)
+
+    in (ValidatedBlock block events rest, idx')
+
+-- | Check whether the given transaction can be validated in the given slot.
+canValidateNow :: Slot -> Tx -> Bool
+canValidateNow slot tx = Interval.member slot (txValidRange tx)
+
+mkEvent :: Tx -> Maybe Index.ValidationError -> NodeClientEvent
+mkEvent t result =
+    case result of
+        Nothing  -> TxnValidate (hashTx t)
+        Just err -> TxnValidationFail (hashTx t) err
+
+-- | Validate a transaction in the current emulator state.
+validateEm :: MonadState Index.UtxoIndex m => Slot -> Tx -> m (Maybe Index.ValidationError)
+validateEm h txn = do
+    idx <- get
+    let result = Index.runValidation (Index.validateTransaction h txn) idx
+    case result of
+        Left e -> pure (Just e)
+        Right idx' -> do
+            _ <- put idx'
+            pure Nothing
+
+class MonadState NodeClientState m => NodeClientAPI m where
+    publishTx :: Tx -> m ()
+    currentSlot :: m Slot
+    eventStream :: Slot -> ConduitT () Block m ()
+    processBlock :: m ValidatedBlock
+            
+-- | Pure implementation of the NodeClient API
+
+instance NodeClientAPI (State NodeClientState) where 
+    publishTx tx = state $ ((), ) . over txPool (tx :)
+
+    currentSlot = 
+        (Slot . fromIntegral . length . _chainNewestFirst) <$> get
+
+    eventStream slot = loop slot
+        where
+            loop :: MonadState NodeClientState m
+                 => Slot 
+                 -> ConduitT () Block m ()
+            loop s = do
+                blocks <-
+                    drop (fromIntegral s) . _chainNewestFirst <$> get
+                void $ yieldMany blocks
+                loop $ s + fromIntegral (length blocks)
```

Comment:
This seems like it's going to busy-loop constantly accessing the state. We probably need some better way of making a "good" producer of blocks.

Reply:
It is commented for now, but streams are generally infinite, and they do look like a busy loop. Here is an example:
```
repeatWhileMC m f =
    loop
  where
    loop = do
        x <- lift m
        when (f x) $ yield x >> loop
```

Reply:
Right, but how "busy" it is depends on e.g. whether `m` returns immediately or might block. For example, if `m` was "poll for data" then it might be fine. But in this case we're not doing anything that blocks, so I'd expect it to maybe peg a CPU or something, which seems bad.